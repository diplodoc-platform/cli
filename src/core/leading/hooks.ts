import type {LeadingPage, Plugin} from './types';

import {AsyncParallelHook, AsyncSeriesHook, AsyncSeriesWaterfallHook} from 'tapable';

import {generateHooksAccess} from '~/core/utils';

export function hooks(name: string) {
    return {
        /**
         * Emits relative to root asset path on each local link in Leading.
         * This includes paths in links and blocks sections.
         */
        Asset: new AsyncParallelHook<[RelativePath, RelativePath]>(
            ['asset', 'path'],
            `${name}.Asset`,
        ),
        Plugins: new AsyncSeriesWaterfallHook<[Plugin[]]>(['plugins'], `${name}.Plugins`),
        Resolved: new AsyncSeriesHook<[DeepFrozen<LeadingPage>, RelativePath]>(
            ['leading', 'path'],
            `${name}.Resolved`,
        ),
        Dump: new AsyncSeriesWaterfallHook<[LeadingPage, NormalizedPath]>(
            ['leading', 'path'],
            `${name}.Dump`,
        ),
    };
}

const [getHooks, withHooks] = generateHooksAccess('Leading', hooks);

export {getHooks, withHooks};
