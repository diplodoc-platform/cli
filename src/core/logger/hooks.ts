import {SyncHook} from 'tapable';

import {generateHooksAccess} from '~/core/utils';

export function hooks(name: string) {
    return {
        Info: new SyncHook<[string]>(['message'], `${name}.Info`),
        Warn: new SyncHook<[string]>(['message'], `${name}.Warn`),
        Error: new SyncHook<[string]>(['message'], `${name}.Error`),
    };
}

const [getHooks, withHooks] = generateHooksAccess('Logger', hooks);

export {getHooks, withHooks};
