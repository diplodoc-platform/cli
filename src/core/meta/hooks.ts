import type {Meta} from './types';

import {AsyncSeriesWaterfallHook} from 'tapable';

import {generateHooksAccess} from '~/core/utils';

export function hooks(name: string) {
    return {
        Dump: new AsyncSeriesWaterfallHook<[Meta, NormalizedPath]>(
            ['meta', 'path'],
            `${name}.Dump`,
        ),
    };
}

const [getHooks, withHooks] = generateHooksAccess('Meta', hooks);

export {getHooks, withHooks};
