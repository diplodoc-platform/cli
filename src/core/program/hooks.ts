import {intercept} from '~/utils';
import type {Command, Config, ExtendedOption} from '~/core/config';
import type {Run} from '~/core/run';
import {AsyncSeriesHook, AsyncSeriesWaterfallHook, SyncHook} from 'tapable';

export function hooks<TConfig extends BaseConfig, TArgs extends BaseArgs>(name: string) {
    return intercept(name, {
        Command: new SyncHook<[Command, ExtendedOption[]]>(
            ['command', 'options'],
            `${name}.Command`,
        ),
        /**
         * Called on resolved config, before it will be merged with args.
         * Best place to validate that config doesn't store some secret data.
         * Config can't be modified here.
         */
        RawConfig: new AsyncSeriesHook<[Config<TConfig>, TArgs]>(
            ['config', 'args'],
            `${name}.Config`,
        ),
        /**
         * Called on resolved config, after it was merged with args.
         * Best place to normalize final config data.
         * Config can be modified here.
         */
        Config: new AsyncSeriesWaterfallHook<[Config<TConfig>, TArgs]>(
            ['config', 'args'],
            `${name}.Config`,
        ),
        /**
         * Async series hook which runs before start of any Run type.<br/><br/>
         * Args:
         * - run - [Build.Run](./Run.ts) constructed context.<br/>
         * Best place to subscribe on Run services hooks.
         */
        BeforeAnyRun: new AsyncSeriesHook<Run<TConfig>>(['run'], `${name}.BeforeAnyRun`),
        // TODO: decompose handler and describe this hook
        AfterAnyRun: new AsyncSeriesHook<Run<TConfig>>(['run'], `${name}.AfterAnyRun`),
    });
}

export const Hooks = Symbol(`BaseHooks`);

export function getHooks<TConfig, TArgs>(program: {
    [Hooks]?: ReturnType<typeof hooks<TConfig, TArgs>>;
}) {
    return program[Hooks] || hooks('Unknown');
}
