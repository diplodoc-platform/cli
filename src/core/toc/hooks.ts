import type {IncludeInfo, IncluderOptions, RawToc, RawTocItem, Toc} from './types';

import {AsyncParallelHook, AsyncSeriesWaterfallHook, HookMap} from 'tapable';

import {generateHooksAccess, intercept} from '~/core/utils';

const name = 'Toc';

export function hooks() {
    return intercept(name, {
        /**
         * Called before item data processing (but after data interpolation)
         */
        Item: new AsyncSeriesWaterfallHook<[RawTocItem, RelativePath]>(
            ['TocItem', 'TocPath'],
            `${name}.Item`,
        ),
        /**
         * AsyncSeriesWaterfall HookMap called for each includer name detected in toc.<br>
         * Expects RawToc as result of waterfall.
         */
        Includer: new HookMap(
            (type: string) =>
                new AsyncSeriesWaterfallHook<[RawToc, IncluderOptions, RelativePath]>(
                    ['Toc', 'options', 'TocPath'],
                    `${name}.Includer(${type})`,
                ),
        ),
        Resolved: new AsyncParallelHook<[Toc, RelativePath]>(
            ['Toc', 'TocPath'],
            `${name}.Resolved`,
        ),
        Included: new AsyncParallelHook<[Toc, RelativePath, IncludeInfo]>(
            ['Toc', 'TocPath', 'IncludeInfo'],
            `${name}.Included`,
        ),
    });
}

const [getHooks, withHooks] = generateHooksAccess(name, hooks);

export {getHooks, withHooks};
