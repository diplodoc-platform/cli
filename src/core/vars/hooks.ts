import type {Preset, Presets} from './types';

import {AsyncParallelHook, AsyncSeriesWaterfallHook} from 'tapable';

import {intercept} from '~/core/utils';

const name = 'Vars';

export function hooks() {
    return intercept(name, {
        /**
         * Async waterfall hook.<br>
         * Called after any presets.yaml was loaded.
         */
        PresetsLoaded: new AsyncSeriesWaterfallHook<[Presets, RelativePath]>(
            ['presets', 'path'],
            `${name}.PresetsLoaded`,
        ),
        /**
         * Async parallel hook.<br>
         * Called after vars was resolved on any level.<br>
         * Vars data is sealed here.
         */
        Resolved: new AsyncParallelHook<[Preset, RelativePath]>(
            ['vars', 'path'],
            `${name}.Resolved`,
        ),
    });
}

export const Hooks = Symbol(`${name}Hooks`);

export function getHooks(program: {[Hooks]?: ReturnType<typeof hooks>} | undefined) {
    return (program && program[Hooks]) || hooks();
}
