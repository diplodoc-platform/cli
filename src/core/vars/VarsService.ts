import type {Presets} from './types';

import {dirname, join} from 'node:path';
import {merge} from 'lodash';
import {dump, load} from 'js-yaml';

import {Run} from '~/commands/build';
import {freeze, normalizePath, own} from '~/utils';

import {Hooks, hooks} from './hooks';

export type VarsServiceConfig = {
    varsPreset: string;
    vars: Hash;
    ignore: string[];
};

export class VarsService {
    [Hooks] = hooks();

    get entries() {
        return [...Object.entries(this.cache)];
    }

    private run: Run;

    private logger: Run['logger'];

    private config: VarsServiceConfig;

    private cache: Record<NormalizedPath, Hash> = {};

    constructor(run: Run) {
        this.run = run;
        this.logger = run.logger;
        this.config = run.config;
    }

    async init() {
        const presets = await this.run.glob('**/presets.yaml', {
            cwd: this.run.input,
        });

        for (const preset of presets) {
            await this.load(preset);
        }
    }

    async load(path: RelativePath) {
        path = normalizePath(path);

        const varsPreset = this.config.varsPreset || 'default';
        const file = normalizePath(join(dirname(path), 'presets.yaml'));

        if (this.cache[file]) {
            return this.cache[file];
        }

        this.logger.proc(file);

        const scopes = [];

        if (dirname(path) !== '.') {
            scopes.push(await this.load(dirname(path)));
        }

        try {
            const presets = await this[Hooks].PresetsLoaded.promise(
                load(await this.run.read(join(this.run.input, file))) as Presets,
                file,
            );

            scopes.push(presets['default']);

            if (varsPreset && varsPreset !== 'default') {
                scopes.push(presets[varsPreset] || {});
            }
        } catch (error) {
            if (!own(error, 'code') || error.code !== 'ENOENT') {
                throw error;
            }
        }

        scopes.push(this.config.vars);

        this.cache[file] = freeze(merge({}, ...scopes));

        await this[Hooks].Resolved.promise(this.cache[file], file);

        return this.cache[file];
    }

    dump(presets: Hash): string {
        return dump(presets, {
            lineWidth: 120,
        });
    }
}
