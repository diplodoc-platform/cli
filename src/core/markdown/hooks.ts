import type {AdditionalInfo, Collect, Plugin} from './types';

import {AsyncSeriesHook, AsyncSeriesWaterfallHook} from 'tapable';

import {generateHooksAccess} from '~/core/utils';
import {Meta} from '~/core/meta';

export function hooks(name: string) {
    return {
        Collects: new AsyncSeriesWaterfallHook<[Collect[]]>(['collects'], `${name}.Collects`),
        Plugins: new AsyncSeriesWaterfallHook<[Plugin[]]>(['plugins'], `${name}.Plugins`),
        Loaded: new AsyncSeriesHook<[string, DeepFrozen<Meta>, NormalizedPath]>(
            ['markdown', 'meta', 'path'],
            `${name}.Loaded`,
        ),
        Resolved: new AsyncSeriesHook<[string, NormalizedPath]>(
            ['markdown', 'path'],
            `${name}.Resolved`,
        ),
        Dump: new AsyncSeriesWaterfallHook<[string, NormalizedPath, AdditionalInfo]>(
            ['markdown', 'path', 'info'],
            `${name}.Dump`,
        ),
    };
}

const [getHooks, withHooks] = generateHooksAccess('Markdown', hooks);

export {getHooks, withHooks};
