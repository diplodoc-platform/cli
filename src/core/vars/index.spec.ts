import type {VarsServiceConfig} from './VarsService';

import {join} from 'node:path';
import {describe, expect, it, vi} from 'vitest';
import {when} from 'vitest-when';
import {dedent} from 'ts-dedent';
import {YAMLException} from 'js-yaml';

import {setupRun} from '~/commands/build/__tests__';

import {getHooks} from './hooks';
import {VarsService} from './VarsService';

const ENOENT = Object.assign(new Error('ENOENT: no such file or directory'), {
    code: 'ENOENT',
});

type Options = Partial<VarsServiceConfig>;

function prepare(content: string | Error | Hash<string | Error>, options: Options = {}) {
    const run = setupRun({
        varsPreset: options.varsPreset,
        vars: options.vars || {},
    });

    const service = new VarsService(run);

    if (typeof content === 'string' || content instanceof Error) {
        content = {'./presets.yaml': content};
    }

    for (const [file, data] of Object.entries(content)) {
        if (data instanceof Error) {
            when(run.read).calledWith(join(run.input, file)).thenReject(data);
        } else {
            when(run.read).calledWith(join(run.input, file)).thenResolve(data);
        }
    }

    return service;
}

async function call(content: string | Error, options: Options = {}) {
    const service = prepare(content, options);
    const result = await service.load('presets.yaml' as NormalizedPath);

    expect(service.dump(result)).toMatchSnapshot();
}

function test(name: string, content: string | Error, options: Options = {}) {
    it(name, async () => call(content, options));
}

describe('vars', () => {
    describe('service', () => {
        describe('load', () => {
            test(
                'should load presets file default scope',
                dedent`
                    default:
                      field1: value1
                      field2: value2
                    internal:
                      field1: value3
                    external:
                      field1: value4
                `,
            );

            test(
                'should load presets file target scope',
                dedent`
                    default:
                      field1: value1
                      field2: value2
                    internal:
                      field1: value3
                    external:
                      field1: value4
                `,
                {varsPreset: 'internal'},
            );

            test(
                'should override default presets with vars',
                dedent`
                    default:
                      field1: value1
                      field2: value2
                    internal:
                      field1: value3
                    external:
                      field1: value4
                `,
                {vars: {field1: 'value6'}},
            );

            test(
                'should override target presets with vars',
                dedent`
                    default:
                      field1: value1
                      field2: value2
                    internal:
                      field1: value3
                    external:
                      field1: value4
                `,
                {varsPreset: 'internal', vars: {field1: 'value6'}},
            );

            test('should use vars if presets not found', ENOENT, {vars: {field1: 'value6'}});

            it('should throw parse error', async () => {
                await expect(() => call('!@#', {vars: {field1: 'value6'}})).rejects.toThrow(
                    YAMLException,
                );
            });

            it('should load super layers', async () => {
                const service = prepare(
                    {
                        './presets.yaml': dedent`
                            default:
                              field1: value1
                              override1: value2
                              override2: value2
                              override3: value2
                              override4: value2
                            internal:
                              field2: value1
                              override1: value1
                        `,
                        './subfolder/presets.yaml': dedent`
                            default:
                              sub1: value1
                              sub2: value2
                              override2: value1
                              override5: value2
                            internal:
                              sub2: value1
                              override3: value1
                              override6: value2
                        `,
                        './subfolder/subfolder/presets.yaml': ENOENT,
                        './subfolder/subfolder/subfolder/presets.yaml': dedent`
                            default:
                              subsub1: value2
                              override4: value2
                              override5: value1
                            internal:
                              subsub1: value1
                              subsub2: value1
                              override4: value1
                              override6: value1
                        `,
                    },
                    {varsPreset: 'internal'},
                );

                const result = await service.load(
                    'subfolder/subfolder/subfolder/presets.yaml' as NormalizedPath,
                );

                expect(service.dump(result)).toMatchSnapshot();
            });

            it('should call PresetsLoaded hook', async () => {
                const service = prepare(dedent`
                    default:
                      field1: value1
                `);

                const spy = vi.fn();

                getHooks(service).PresetsLoaded.tap('Test', spy);

                await service.load('presets.yaml' as NormalizedPath);

                expect(spy).toHaveBeenCalledWith({default: {field1: 'value1'}}, 'presets.yaml');
            });

            it('should call Resolved hook', async () => {
                const service = prepare(dedent`
                    default:
                      field1: value1
                `);

                const spy = vi.fn();

                getHooks(service).Resolved.tap('Test', spy);

                await service.load('presets.yaml' as NormalizedPath);

                expect(spy).toHaveBeenCalledWith({field1: 'value1'}, 'presets.yaml');
            });

            it('should allow content updating in PresetsLoaded hook', async () => {
                const service = prepare(dedent`
                    default:
                      field1: value1
                `);

                getHooks(service).PresetsLoaded.tap('Test', (presets) => {
                    presets.default.field1 = 'value2';

                    return presets;
                });

                const result = await service.load('presets.yaml' as NormalizedPath);

                expect(service.dump(result)).toMatchSnapshot();
            });

            it('should allow content extending in PresetsLoaded hook', async () => {
                const service = prepare(dedent`
                    default:
                      field1: value1
                `);

                getHooks(service).PresetsLoaded.tap('Test', (presets) => {
                    presets.default.field2 = 'value2';

                    return presets;
                });

                const result = await service.load('presets.yaml' as NormalizedPath);

                expect(service.dump(result)).toMatchSnapshot();
            });

            it('should reject content updating in Resolved hook', async () => {
                const service = prepare(dedent`
                    default:
                      field1: value1
                `);

                getHooks(service).Resolved.tap('Test', (vars) => {
                    vars.field1 = 'value2';
                });

                await expect(() =>
                    service.load('presets.yaml' as NormalizedPath),
                ).rejects.toThrow();
            });

            it('should reject content extending in Resolved hook', async () => {
                const service = prepare(dedent`
                    default:
                      field1: value1
                `);

                getHooks(service).Resolved.tap('Test', (vars) => {
                    vars.field2 = 'value2';
                });

                await expect(() =>
                    service.load('presets.yaml' as NormalizedPath),
                ).rejects.toThrow();
            });

            it('should load content only once', async () => {
                const service = prepare(dedent`
                    default:
                      field1: value1
                `);

                const spy1 = vi.fn();
                const spy2 = vi.fn();

                getHooks(service).PresetsLoaded.tap('Test', spy1);
                getHooks(service).Resolved.tap('Test', spy2);

                await service.load('presets.yaml' as NormalizedPath);
                await service.load('presets.yaml' as NormalizedPath);

                expect(spy1).toHaveBeenCalledOnce();
                expect(spy2).toHaveBeenCalledOnce();
            });
        });
    });
});
