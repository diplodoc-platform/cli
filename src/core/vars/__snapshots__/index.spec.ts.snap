// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`vars > service > load > should handle empty preset 1`] = `
"field1: value1
override1: value2
override2: value2
override3: value2
override4: value2
"
`;

exports[`vars > service > load > should load presets file default scope 1`] = `
"field1: value1
override1: value2
override2: value2
override3: value2
override4: value2
"
`;

exports[`vars > service > load > should load presets file target scope 1`] = `
"field2: value1
override1: value1
field1: value1
override2: value2
override3: value2
override4: value2
"
`;

exports[`vars > service > load > should merge deep presets 1 1`] = `
"subsub1: value2
override4: value2
override5: value1
sub1: value1
sub2: value2
override2: value1
field1: value1
override1: value2
override3: value2
"
`;

exports[`vars > service > load > should merge deep presets 2 1`] = `
"subsub1: value2
override4: value2
override5: value1
sub1: value1
sub2: value2
override2: value1
field1: value1
override1: value2
override3: value2
"
`;

exports[`vars > service > load > should not merge presets on lower levels 1`] = `
"sub1: value1
sub2: value2
override2: value1
override5: value2
field1: value1
override1: value2
override3: value2
override4: value2
"
`;

exports[`vars > service > load > should override default presets with vars 1`] = `
"field1: value6
override1: value2
override2: value2
override3: value2
override4: value2
"
`;

exports[`vars > service > load > should override target presets with vars 1`] = `
"field1: value6
field2: value1
override1: value1
override2: value2
override3: value2
override4: value2
"
`;
